src/main/java/com/alipay/oceanbase/rpc/location/LocationUtil.java

```diff
@@ -742,53 +742,73 @@ public static TableEntry getTableEntryLocationFromRemote(Connection connection,
+                    sb.append(ObPartIdCalculator.generatePartId(firstPartId, subPartId));
                 }
             } else {
-                for (int i = 0; i < partitionNum; i++) {
-                    if (i > 0) {
+                for (int i = offset; i < endOffset; i++) {
+                    if (i > offset) {
                         sb.append(", ");
                     }
                     sb.append(i);
                 }
             }
             sql = MessageFormat.format(PROXY_LOCATION_SQL_PARTITION, sb.toString());
         }
-        try {
-            ps = connection.prepareStatement(sql);
-            ps.setString(1, key.getTenantName());
-            ps.setString(2, key.getDatabaseName());
-            ps.setString(3, key.getTableName());
+        return sql;
+    }
 
-            rs = ps.executeQuery();
-            partitionEntry = getPartitionLocationFromResultSet(tableEntry, rs);
-            tableEntry.setPartitionEntry(partitionEntry);
-            tableEntry.setRefreshTimeMills(System.currentTimeMillis());
-        } catch (Exception e) {
-            RUNTIME.error(LCD.convert("01-00010"), key, partitionNum, tableEntry, e);
-            throw new ObTablePartitionLocationRefreshException(
-                format(
-                    "fail to get partition location entry from remote entryKey = %s partNum = %d tableEntry =%s",
-                    key, partitionNum, tableEntry), e);
-        } finally {
+    /*
+     * Get table entry location from remote.
+     */
+    public static TableEntry getTableEntryLocationFromRemote(Connection connection,
+                                                             TableEntryKey key,
+                                                             TableEntry tableEntry)
+                                                                                   throws ObTablePartitionLocationRefreshException {
+
+        PreparedStatement ps = null;
+        ResultSet rs = null;
+        ObPartitionEntry partitionEntry = new ObPartitionEntry();
+        long partitionNum = tableEntry.getPartitionNum();
+        int epoch = (int) ((partitionNum / MAX_TABLET_NUMS_EPOCH) + 1);
+        for (int i = 0; i < epoch; i++) {
             try {
-                if (null != rs) {
-                    rs.close();
-                }
-                if (null != ps) {
-                    ps.close();
+                int offset = i * MAX_TABLET_NUMS_EPOCH;
+                String sql = genLocationSQLByOffset(tableEntry, offset, MAX_TABLET_NUMS_EPOCH);
+                ps = connection.prepareStatement(sql);
+                ps.setString(1, key.getTenantName());
+                ps.setString(2, key.getDatabaseName());
+                ps.setString(3, key.getTableName());
+
+                rs = ps.executeQuery();
+                partitionEntry = getPartitionLocationFromResultSet(tableEntry, rs, partitionEntry);
+                tableEntry.setPartitionEntry(partitionEntry);
```

Comment:
Should we set the `partitionEntry` after all the info is collected?

Reply:
fix